end-to-end-tests/echo_test.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	payload              = "0xdeadbeef"
+	maxReadInputAttempts = 10
+	defaulTxDatabaseFile = "default_tx_database"
+)
+
+type EchoInputTestSuite struct {
+	suite.Suite
+	containers    *deps.DepsContainers
+	cancel        context.CancelFunc
+	tempDir       string
+	supervisorErr chan error
+}
+
+func (s *EchoInputTestSuite) SetupSuite() {
+
+	// Clear default_tx_database
+	cwd, err := os.Getwd()
+	require.NoError(s.T(), err)
+
+	_ = os.Remove(filepath.Join(cwd, defaulTxDatabaseFile))
+
+	// Create machine snapshot
+	tempDir, err := ioutil.TempDir("", "machine-snapshot")
+	assert.NoError(s.T(), err)
+	machine.Save("cartesi/rollups-node-snapshot:devel", tempDir, "test-echo-app")
+
+	// Run deps
+	ctx := context.Background()
+	var depsConfig = deps.NewDefaultDepsConfig()
+	depsContainers, err := deps.Run(ctx, *depsConfig)
+	require.NoError(s.T(), err)
+
+	// Wait a little for Dependencies to be actually ready
+	time.Sleep(2 * time.Second)
+
+	// Fix the Blochain timestamp. Must be "in the future"
+	err = SetNextBlockTimestamp(ctx, LocalBlockchainHttpEndpoint, 7000000000)
+	require.NoError(s.T(), err)
+
+	// Run Node Service
+	nodeConfig := NewLocalNodeConfig()
+
+	nodeConfig.SnapshotDir = tempDir
+
+	ctx, cancel := context.WithCancel(ctx)
+	supervisor, err := node.Setup(ctx, nodeConfig)
+	require.NoError(s.T(), err)
+
+	ready := make(chan struct{}, 1)
+	supervisorErr := make(chan error, 1)
+	go func() {
+		err := supervisor.Start(ctx, ready)
+		if err != nil {
+			supervisorErr <- err
+		}
+	}()
+
+	select {
+	case err := <-supervisorErr:
+		require.NoError(s.T(), err)
+	case <-ready:
+		break
+	}
+
+	// Configure Suite for tear down
+	s.containers = depsContainers
+	s.tempDir = tempDir
+	s.cancel = cancel
+	s.supervisorErr = supervisorErr
+
+}
+
+func (s *EchoInputTestSuite) TearDownSuite() {
+
+	// Stop Node services
+	s.cancel()
+
+	// Remove machine snpshot
+	os.RemoveAll(s.tempDir)
+
+	// Terminate deps
+	ctx := context.Background()
+	err := deps.Terminate(ctx, s.containers)
+	require.NoError(s.T(), err)
+
+	// Clear default_tx_database
+	cwd, err := os.Getwd()
+	require.NoError(s.T(), err)
+
+	_ = os.Remove(filepath.Join(cwd, defaulTxDatabaseFile))
+
+}
+
+func (s *EchoInputTestSuite) SetupTest() {
+
+}
+
+func (s *EchoInputTestSuite) TestSendInput() {
+
+	ctx := context.Background()
+
+	inputIndex, err := AddInput(ctx, LocalBlockchainHttpEndpoint, payload)
+	require.NoError(s.T(), err)
+
+	// Check input was correctly added to the blockchain
+	require.EqualValues(s.T(), 0, inputIndex)
+
+	require.NoError(s.T(), AdvanceTime(ctx, LocalBlockchainHttpEndpoint, 120))
+
+	// Get Input with vouchers and proofs
+	graphQlClient := graphql.NewClient("http://localhost:10000/graphql", nil)
+	getInputChan := make(chan *readerclient.Input, 1)
+	getInputErr := make(chan struct{}, 1)
+
+	go func() {
```

Comment:
I didn't understand why this needed to be a goroutine.

Reply:
The idea was that the following `select` would wait for what happens first : input being correctly retrieved or supervisor giving an error. 

Even if the node died, the loop will continue until it reaches the `maxReadInputAttemps`. 

But as it does not takes that long to execute, the necessity of being a goroutine is very low.